### DISCUSSION TITLE
Clarification/advice on primeNG theming

### DISCUSSION THREAD

**Discussion by VasileiosP** [primefaces/community#148] [PrimeNG] [answered]
Hello!

We would like some clarifications and advice on our future implementation regarding primeNG theming.

Our current implementation is the following:

We have created a common library of reusable components, based on primeNG, that is imported and used by our 5 angular applications. Essentially, each of the implemented components acts as a wrapper for a PrimeNG component, with custom styling applied to align with our specific UX requirements Every application that imports the common library, also imports 'primeng.min.css', 'primeicons.css' and 'saga-blue/theme.css'. So basically our css extra styling overrides the 'saga-blue' prime theme.

However, we have recognized that this approach no longer adequately addresses our evolving needs, especially concerning the implementation of a dark theme across our applications.

In light of this, we have revisited the official documentation and made the decision to adopt the recommended approach using the [primeng-sass-theme repository](https://github.com/primefaces/primeng-sass-theme). 
With this approach, we will be able to create custom themes(both light and dark) following the official guidelines as described here from 6:00: https://www.youtube.com/watch?v=JAapJf6q1cw. This will also give us the flexibility to introduce additional overrides where required. (we need to define some extra colors with variations and establish global CSS variables for all applications to use.)

Our plan is to integrate these custom theme files into the common library project so as to leverage the advantages of build-time compilation and quick testing/debugging.
When theming is ready, the 2 css files will be compiled manually and imported by the angular apps where they can switch light and dark css files on runtime.

Our current angular and primeNG version is 15, it is planned to upgrade to 16 in October. (we will probably first upgrade and then start implementing the new themes).

We would like your expert advice on the following:

1. Is this the best approach to follow?
2. Is this implementation going to be affected by the refactoring you have planned as announced here in the home page note (Note: In upcoming versions, theming architecture will be redesigned to utilize CSS variables instead of SCSS variables in a backward compatible way for a dynamic approach. In addition, a new Unstyled mode will be provided as an alternative to the default styling so that CSS libraries like Tailwind or Bootstrap can be used to style the components. This work is planned to be completed in Q3 2023.)
3. Should we wait until this refactoring is finished?
4. Is your theming architecture redesign going to be completed in Q3?
5. Is this project([primeng-sass-theme repository](https://github.com/primefaces/primeng-sass-theme)) going to support future angular versions as it does until now?

Thank you!

**Comment by cetincakiroglu** [comment, marked as answer]
Hi,

Thanks for using PrimeNG in your apps.

1-) Yes it's recommended approach. We advice you should use either latest versions or LTS versions of PrimeNG along with primeng sass theme to get the latest updates, bug fixes and theme updates.
2-) We have plans to implement unstyled mode to PrimeNG and it won't break previous theming, we have implemented it into PrimeReact and PrimeVue, you can check it to have an idea.
3-) You don't need to wait, this updates will not break your apps. In the worst case scenario, we may change some classes in the sass-theme but it would be minor.
4-) Right now we're implementing accessibility to the PrimeNG and we're planning to release it in October. Afterwards we're planning to implement unstyled and passthrough to the library.
5-) primeng-sass-theme and PrimeNG versions are in match.  For example, we suggest using PrimeNG 16.4.1 with the sass-theme 16.4.1 to get the latest style updates. Sass-theme is not bounded with Angular but PrimeNG is (you should use matching versions PrimeNG 16x & Angular 16x)

From now on, you can update your version from 15 to 16. In October we'll deliver the accessibility. Probably Angular will receive a major update (v17) in November and we'll match the PrimeNG version with it too.

Hope it helps.

**Reply by VasileiosP** [reply]
Thank you for your answer, it was very helpful!

**Comment by VasileiosP** [comment]
Hello! As we are planning to start the theming implementation (exactly as described on the first message - using the **primeng-sass-theme** repository), we would like to confirm again some things, as it will affect many projects.

- As we know that you are working on a new dynamic approach with js presets, will the primeng-sass-theme repo keep being updated/supported by you in the future? On angular version upgrades will we need extra work to adjust our themes?
- When you complete the unstyled mode implementation, will it be recommended for us(and relatively easy) to migrate to that approach our theming?
- Do you have any estimation on when this dynamic implementation will be completed?
Thanks a lot
